x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/FunctionRegistry.java

```diff
@@ -91,98 +92,109 @@
 import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.Map.Entry;
 import java.util.TimeZone;
+import java.util.TreeSet;
 import java.util.function.BiFunction;
 import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 import static java.util.Collections.emptyList;
 import static java.util.Collections.unmodifiableList;
 import static java.util.stream.Collectors.toList;
 
 public class FunctionRegistry {
-    private static final List<FunctionDefinition> DEFAULT_FUNCTIONS = unmodifiableList(Arrays.asList(
-        // Aggregate functions
-            def(Avg.class, Avg::new),
-            def(Count.class, Count::new),
-            def(Max.class, Max::new),
-            def(Min.class, Min::new),
-            def(Sum.class, Sum::new),
+    // list of functions grouped by type of functions (aggregate, statistics, math etc) and ordered alphabetically inside each group
+    private static final List<TreeSet<FunctionDefinition>> DEFAULT_FUNCTIONS = unmodifiableList(Arrays.asList(
```

Comment:
The `TreeSet` is not needed on the function list itself - it can only be used internally to sort a group/batch and then simply inserted in the list (which preserves ordering).

Reply:
The `DEFAULT_FUNCTIONS` is used afterwards to build the `defs` LinkedHashMap which, in turn, is used for `SHOW FUNCTIONS` output. If, at the time of the map build, there is no formal grouping (List, Tree, Map etc) of Math/Aggregations/String/Date functions, then I think it would be difficult to sort the groups of functions considering also the aliases (the `defs` map is extended and also includes the aliases, not only the functions themselves and each group of functions is sorted individually).

You are right that the list should contain only the functions, and the `List<TreeSet<FunctionDefintion>>` structure is too complex for what the list should actually represent. Since it's only used internally to build the `defs` HashMap, I will refactor the DEFAULT_FUNCTIONS into a method to be called by the constructor, as you suggested.